chainstate/tx-verifier/src/transaction_verifier/input_output_policy/constraints_accumulator.rs

```diff
@@ -78,640 +112,440 @@ impl ConstrainedValueAccumulator {
-        let mut rng = make_seedable_rng(seed);
-
-        let chain_config = common::chain::config::Builder::new(ChainType::Mainnet)
-            .consensus_upgrades(NetUpgrades::regtest_with_pos())
-            .build();
-        let required_maturity_distance =
-            chain_config.staking_pool_spend_maturity_distance(BlockHeight::new(1));
-
-        let pool_id = PoolId::new(H256::zero());
-        let staked_atoms = rng.gen_range(100..1000);
-        let fee_atoms = rng.gen_range(1..100);
-        let stake_pool_data = create_stake_pool_data(&mut rng, staked_atoms);
-
-        let pledge_getter = |_| Ok(Some(Amount::from_atoms(staked_atoms)));
-
-        let inputs = vec![TxInput::Utxo(UtxoOutPoint::new(
-            OutPointSourceId::BlockReward(Id::new(H256::random_using(&mut rng))),
-            0,
-        ))];
-        let input_utxos = vec![Some(TxOutput::CreateStakePool(
-            pool_id,
-            Box::new(stake_pool_data),
-        ))];
-
-        let outputs = vec![TxOutput::LockThenTransfer(
-            OutputValue::Coin(Amount::from_atoms(staked_atoms - fee_atoms)),
-            Destination::AnyoneCanSpend,
-            OutputTimeLock::ForBlockCount(required_maturity_distance.to_int() as u64),
-        )];
-
-        let mut constraints_accumulator = ConstrainedValueAccumulator::new();
-
-        constraints_accumulator
-            .process_inputs(
-                &chain_config,
-                BlockHeight::new(1),
-                pledge_getter,
-                &inputs,
-                &input_utxos,
-            )
-            .unwrap();
-
-        constraints_accumulator.process_outputs(&chain_config, &outputs).unwrap();
-
-        assert_eq!(
-            constraints_accumulator.consume().unwrap().into_atoms(),
-            fee_atoms
-        );
-    }
+    fn process_output_timelock(
+        &mut self,
+        timelock: &OutputTimeLock,
+        locked_coins: Amount,
+    ) -> Result<(), IOPolicyError> {
+        match timelock {
+            OutputTimeLock::UntilHeight(_)
+            | OutputTimeLock::UntilTime(_)
+            | OutputTimeLock::ForSeconds(_) => {
+                decrease_or(
+                    &mut self.unconstrained_value,
+                    CoinOrTokenId::Coin,
+                    locked_coins,
+                    IOPolicyError::AttemptToPrintMoneyOrViolateTimelockConstraints(
+                        CoinOrTokenId::Coin,
+                    ),
+                )?;
+            }
+            OutputTimeLock::ForBlockCount(block_count) => {
+                // find the range that can be satisfied with the current timelock
+                let mut constraint_range_iter = self
+                    .timelock_constrained
+                    .range_mut((
+                        std::ops::Bound::Unbounded,
+                        std::ops::Bound::Included(block_count),
+                    ))
+                    .rev()
+                    .peekable();
```

Comment:
Looks like you could "prepend" (using `.chain`) the unconstrained coins amount and remove the `None` case  in the loop below.

Reply:
I like the idea, but couldn't implement it nicely. The problem is that unconstrained coins are optional so I resulted having conflicting types for the iterator.

Reply:
> I like the idea, but couldn't implement it nicely. The problem is that unconstrained coins are optional so I resulted having conflicting types for the iterator.

Not sure it helps but you can get an iterator from `Option` (it returns zero or one element).

Reply:
Gave it another try and it worked out, thanks